### PR TITLE
Create pre-validations for the RegisterBtcTransfer function

### PR DESCRIPTION
**Pull request by pamgonzalez**
This commit is only based on creating the previous validations for this function. It does not include the logic of the functionality itself.

### CODE REVIEW

File: rskj-core/src/main/java/co/rsk/peg/BridgeSupport.java [hunk 1]
```diff
@@ -2073,11 +2073,69 @@ public int registerBtcTransfer(
             Address userRefundAddress,
             RskAddress lbcAddress,
             Address lpbtcAddress,
-            boolean executionStatus
-    ) {
+            boolean shouldTransferToContract
+    ) throws BlockStoreException, RegisterBtcTransferException, IOException {
+        if (!BridgeUtils.isContractTx(rskTx)) {
+            String errorMesage = String.format("[registerBtcTransfer] [rskTx:%s] Transaction not a contract", ByteUtil.toHexString(rskTx.getHash().getBytes()));
```
Comment (guidohernan93) [inline rskj-core/src/main/java/co/rsk/peg/BridgeSupport.java:2079]: Variable should be `errorMessage`

Comment (guidohernan93) [reply rskj-core/src/main/java/co/rsk/peg/BridgeSupport.java:2079]: Line is too long, add new line

File: rskj-core/src/main/java/co/rsk/peg/BridgeSupport.java [hunk 2]
```diff
@@ -2073,11 +2073,69 @@ public int registerBtcTransfer(
...
+            throw new RegisterBtcTransferContractValidationException(errorMesage);
+        }
+
+        BtcTransaction btcTx = new BtcTransaction(bridgeConstants.getBtcParams(), btcTxSerialized);
+        Coin totalAmount = this.getAmountToActiveFederation(btcTx);
+
+        Sha256Hash btcTxHash = BtcTransactionFormatUtils.calculateBtcTxHash(btcTxSerialized);
+
+        if (!validationsForRegisterBtcTransaction(btcTxHash, height, pmtSerialized, btcTxSerialized)) {
+            String errorMesage = String.format("[registerBtcTransfer] [rskTx:%s] error during validationsForRegisterBtcTransaction", ByteUtil.toHexString(btcTxHash.getBytes()));
```
Comment (guidohernan93) [inline rskj-core/src/main/java/co/rsk/peg/BridgeSupport.java:2090]: Same typo, variable should be`errorMessage`, and should add new line

File: rskj-core/src/main/java/co/rsk/peg/BridgeSupport.java [hunk 3]
```diff
@@ -2073,11 +2073,69 @@ public int registerBtcTransfer(
...
+            logger.warn(errorMesage);
+            throw new RegisterBtcTransferRegisterBtcTransValidationException(errorMesage);
+        }
+
+        if (!provider.isFastBridgeFederationDerivationHashUsed(derivationArgumentsHash)) {
+            logger.warn("[registerBtcTransfer] [btcTxHash:{}] derivationArgumentsHash is already saved in BridgeStorageProvider ", btcTxHash);
```
Comment (guidohernan93) [inline rskj-core/src/main/java/co/rsk/peg/BridgeSupport.java:2096]: Condition here should be true, we are checking if derivation arguments hash was already used

File: rskj-core/src/main/java/co/rsk/peg/BridgeSupport.java [hunk 4]
```diff
@@ -2073,11 +2073,69 @@ public int registerBtcTransfer(
...
+                totalAmount
+        )) {
+            generateRejectionRelease(btcTx, lpbtcAddress, rskTx, totalAmount, true);
+            return -2;
+        }
+
         return 1;  //TODO: Includes logic
     }
 
+    private Address getSenderBtcAddress(BtcTransaction btcTx) throws GetSenderBtcAddressSenderNotPresentException {
```
Comment (guidohernan93) [inline rskj-core/src/main/java/co/rsk/peg/BridgeSupport.java:2119]: `getSenderBtcAddress` & `getAmountToActiveFederation` are duplicating existing code on `processPegin`. That method should be invoked from `processPegin`.

File: rskj-core/src/main/java/co/rsk/peg/GetSenderBtcAddressSenderNotPresentException.java [hunk 5]
```diff
@@ -0,0 +1,7 @@
+package co.rsk.peg;
+
+public class GetSenderBtcAddressSenderNotPresentException extends RegisterBtcTransferException{
+    public GetSenderBtcAddressSenderNotPresentException(String message) {
+        super(message);
+    }
+}
```
Comment (guidohernan93) [inline rskj-core/src/main/java/co/rsk/peg/GetSenderBtcAddressSenderNotPresentException.java:7]: Missing new line.

File: rskj-core/src/main/java/co/rsk/peg/RegisterBtcTransferContractValidationException.java [hunk 6]
```diff
@@ -0,0 +1,7 @@
+package co.rsk.peg;
+
+public class RegisterBtcTransferContractValidationException extends RegisterBtcTransferException {
+    public RegisterBtcTransferContractValidationException(String message) {
+        super(message);
+    }
+}
```
Comment (guidohernan93) [inline rskj-core/src/main/java/co/rsk/peg/RegisterBtcTransferContractValidationException.java:7]: Missing new line.

File: rskj-core/src/main/java/co/rsk/peg/RegisterBtcTransferException.java [hunk 7]
```diff
@@ -0,0 +1,7 @@
+package co.rsk.peg;
+
+public class RegisterBtcTransferException extends Exception {
+    public RegisterBtcTransferException(String message) {
+        super(message);
+    }
+}
```
Comment (guidohernan93) [inline rskj-core/src/main/java/co/rsk/peg/RegisterBtcTransferException.java:7]: Missing new line.

File: rskj-core/src/main/java/co/rsk/peg/RegisterBtcTransferRegisterBtcTransValidationException.java [hunk 8]
```diff
@@ -0,0 +1,7 @@
+package co.rsk.peg;
+
+public class RegisterBtcTransferRegisterBtcTransValidationException extends RegisterBtcTransferException{
+    public RegisterBtcTransferRegisterBtcTransValidationException(String message) {
+        super(message);
+    }
+}
```
Comment (guidohernan93) [inline rskj-core/src/main/java/co/rsk/peg/RegisterBtcTransferRegisterBtcTransValidationException.java:7]: Missing new line.

File: rskj-core/src/main/java/co/rsk/peg/BridgeSupport.java [hunk 9]
```diff
@@ -2073,11 +2073,69 @@ public int registerBtcTransfer(
...
+        Optional<BtcLockSender> btcLockSenderOptional = btcLockSenderProvider.tryGetBtcLockSender(btcTx);
+        if (!btcLockSenderOptional.isPresent() ||
+                !BridgeUtils.txIsProcessable(btcLockSenderOptional.get().getType(), activations)) {
+            logger.warn("[getSenderBtcAddress] [btcTx:{}] Could not get BtcLockSender from Btc tx", btcTx.getHash());
```
Comment (guidohernan93) [inline rskj-core/src/main/java/co/rsk/peg/BridgeSupport.java:2123]: log and exception should use the same string for error, and leave it on a variable

File: rskj-core/src/test/java/co/rsk/peg/BridgeSupportTest.java [hunk 10]
```diff
@@ -4983,6 +4984,275 @@ public void processPegIn_tx_no_lockable_by_surpassing_locking_cap() throws IOExc
...
+                "ffff0502cc000101ffffffff029c070395000000002321036d6b5bc8c0e902f296b5bdf3dfd4b6f095d8d0987818a557e1766" +
+                "ea25c664524ac0000000000000000266a24aa21a9edfeb3b9170ae765cc6586edd67229eaa8bc19f9674d64cb10ee8a205f4c" +
+                "cf0bc60120000000000000000000000000000000000000000000000000000000000000000000000000");
+        byte[]  btcTxSerialized = Hex.decode("020000000001010000000000000000000000000000000000000000000000000000000000000000ffff" +
+                "ffff0502cc000101ffffffff029c070395000000002321036d6b5bc8c0e902f296b5bdf3dfd4b6f095d8d0987818a557e1766" +
+                "ea25c664524ac0000000000000000266a24aa21a9edfeb3b9170ae765cc6586edd67229eaa8bc19f9674d64cb10ee8a205f4c" +
+                "cf0bc60120000000000000000000000000000000000000000000000000000000000000000000000000");
+
+        ActivationConfig.ForBlock activations = mock(ActivationConfig.ForBlock.class);
+        when(activations.isActive(ConsensusRule.RSKIP176)).thenReturn(false);
```
Comment (guidohernan93) [inline rskj-core/src/test/java/co/rsk/peg/BridgeSupportTest.java:5022]: Before RSKIP 176 this method should not be allowed to be called.

File: rskj-core/src/test/java/co/rsk/peg/BridgeSupportTest.java [hunk 11]
```diff
@@ -4983,6 +4984,275 @@ public void processPegIn_tx_no_lockable_by_surpassing_locking_cap() throws IOExc
...
+                mock(Address.class),
+                mock(RskAddress.class),
+                mock(Address.class),
+                true);
+    }
+
+    @Test
+    public void registerBtcTransfer_providerFastBridgeFederationScriptHashReturnNotPresent() throws RegisterBtcTransferException, IOException, BlockStoreException {
+        InternalTransaction rskTx = new InternalTransaction(null, 0, 0, null, null, null, null, null, null, null, null);
+        Keccak256 hash = new Keccak256(HashUtil.keccak256(new byte[]{}));
```
Comment (guidohernan93) [inline rskj-core/src/test/java/co/rsk/peg/BridgeSupportTest.java:5049]: Unused variable `hash`

File: rskj-core/src/test/java/co/rsk/peg/BridgeSupportTest.java [hunk 12]
```diff
@@ -4983,6 +4984,275 @@ public void processPegIn_tx_no_lockable_by_surpassing_locking_cap() throws IOExc
...
+                mock(RskAddress.class),
+                mock(Address.class),
+                true);
+
+        Assert.assertEquals(-1, result);
+    }
+
+    @Test
+    public void registerBtcTransfer_shouldTransferToContractFalse() throws RegisterBtcTransferException, IOException, BlockStoreException {
+        InternalTransaction rskTx = new InternalTransaction(null, 0, 0, null, null, null, null, null, null, null, null);
```
Comment (guidohernan93) [inline rskj-core/src/test/java/co/rsk/peg/BridgeSupportTest.java:5095]: Line too long

File: rskj-core/src/test/java/co/rsk/peg/BridgeSupportTest.java [hunk 13]
```diff
@@ -4983,6 +4984,275 @@ public void processPegIn_tx_no_lockable_by_surpassing_locking_cap() throws IOExc
...
+        int height = 111;
+        byte[]  pmtSerialized = Hex.decode("020000000001010000000000000000000000000000000000000000000000000000000000000000ffff" +
```
Comment (guidohernan93) [inline rskj-core/src/test/java/co/rsk/peg/BridgeSupportTest.java:5097]: I think we should avoid using raw transactions

File: rskj-core/src/main/java/co/rsk/peg/BridgeSupport.java (hunk 4 above)
Comment (guidohernan93) [reply rskj-core/src/main/java/co/rsk/peg/BridgeSupport.java:2119]: I think `getSenderBtcAddress` should be protected too, and both methods should be tested

File: rskj-core/src/test/java/co/rsk/peg/BridgeSupportTest.java [hunk 14]
```diff
@@ -4983,6 +4985,467 @@ public void processPegIn_tx_no_lockable_by_surpassing_locking_cap() throws IOExc
...
+            btcParams,
+            "n3PLxDiwWqa5uH7fSbHCxS6VAjD9Y7Rwkj"
+        );
+
+        BtcTransaction tx = new BtcTransaction(bridgeConstants.getBtcParams());
+        Assert.assertEquals(Collections.emptyList(), bridgeSupport.getUTXOsForAddress(tx, btcAddress));
+    }
+
+    @Test
+    public void getFastBridgeWallet() {
```
Comment (pamgonzalez) [inline rskj-core/src/test/java/co/rsk/peg/BridgeSupportTest.java:5404]: Mark with Ok in the function name if the test is a successful one

File: rskj-core/src/test/java/co/rsk/peg/BridgeSupportTest.java [hunk 15]
```diff
@@ -4983,6 +4985,467 @@ public void processPegIn_tx_no_lockable_by_surpassing_locking_cap() throws IOExc
...
+            lbcAddress,
+            btcAddress,
+            true
+        );
+
+        Assert.assertEquals(1, result);
+    }
+
+    @Test
+    public void getUTXOsForAddress() {
```
Comment (pamgonzalez) [inline rskj-core/src/test/java/co/rsk/peg/BridgeSupportTest.java:5343]: Mark with Ok in the function name if the test is a successful one

File: rskj-core/src/main/java/co/rsk/peg/BridgeSupport.java [hunk 16]
```diff
@@ -2072,12 +2081,203 @@ public int registerBtcTransfer(
...
+        if (!verifyLockDoesNotSurpassLockingCap(btcTx, null, totalAmount)) {
+            WalletProvider walletProvider = createWallet();
+            generateRejectionRelease(btcTx, lpBtcAddress, rskTx, totalAmount, walletProvider);
+            return -2;
+        }
+
         return 1;  //TODO: Includes logic
     }
 
+    protected WalletProvider createWallet() {
```
Comment (pamgonzalez) [inline rskj-core/src/main/java/co/rsk/peg/BridgeSupport.java:2169]: why this function is protected?

File: rskj-core/src/main/java/co/rsk/peg/BridgeSupport.java [hunk 17]
```diff
@@ -2222,8 +2462,8 @@ private boolean verifyLockDoesNotSurpassLockingCap(BtcTransaction btcTx, Address
         return false;
     }
 
-    private Optional<ReleaseTransactionBuilder.BuildResult> getRefundingTransaction(BtcTransaction btcTx, Address senderBtcAddress, boolean isFastBridgeCompatible) throws IOException {
-
+    private Optional<ReleaseTransactionBuilder.BuildResult> getRefundingTransaction(
+        BtcTransaction btcTx, Address refundingBtcAddress, boolean isFastBridgeCompatible) throws IOException {
```
Comment (pamgonzalez) [inline rskj-core/src/main/java/co/rsk/peg/BridgeSupport.java:2466]: I think it does not meet the code standards

File: rskj-core/src/main/java/co/rsk/peg/BridgeSupport.java (hunk 16 above)
Comment (guidohernan93) [reply rskj-core/src/main/java/co/rsk/peg/BridgeSupport.java:2169]: Could be private agree

File: rskj-core/src/main/java/co/rsk/peg/BridgeSupport.java (hunk 17 above)
Comment (guidohernan93) [reply rskj-core/src/main/java/co/rsk/peg/BridgeSupport.java:2466]: This method is no longer nedeed

File: rskj-core/src/main/java/co/rsk/peg/BridgeSupport.java [hunk 18]
```diff
@@ -2072,12 +2068,203 @@ public int registerBtcTransfer(
...
+            return -1;
+        }
+
+        if (!shouldTransferToContract) {
+            logger.warn("[registerBtcTransfer] [btcTxHash:{}] shouldTransferToContract is set as False",
+                btcTxHash
+            );
+            WalletProvider walletProvider = createWallet();
+            generateRejectionRelease(btcTx, userRefundAddress, rskTx, totalAmount, walletProvider);
+            return -1;
```
[the remark this answers is not in the document]
Comment (guidohernan93) [reply rskj-core/src/main/java/co/rsk/peg/BridgeSupport.java:2144]: Will be discussed on https://rsklabs.atlassian.net/browse/RFS-891

**Review by guidohernan93 — APPROVED** [review]

**Comment by pamgonzalez** [comment]
Approved by me

**Pull request merged**